packages/framework/src/Services/BuildService.php

```diff
@@ -0,0 +1,144 @@
+<?php
+
+namespace Hyde\Framework\Services;
+
+use Hyde\Framework\Concerns\InteractsWithDirectories;
+use Hyde\Framework\Hyde;
+use Hyde\Framework\Modules\Routing\RouteContract as Route;
+use Hyde\Framework\Modules\Routing\Router;
+use Hyde\Framework\StaticPageBuilder;
+use Illuminate\Console\Concerns\InteractsWithIO;
+use Illuminate\Console\OutputStyle;
+use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\File;
+
+/**
+ * Moves logic from the build command to a service.
+ *
+ * Handles the build loop which generates the static site.
+ *
+ * @see \Hyde\Framework\Testing\Feature\StaticSiteServiceTest
+ */
+class BuildService
+{
+    use InteractsWithIO;
+    use InteractsWithDirectories;
+
+    protected Router $router;
+
+    public function __construct(OutputStyle $output)
+    {
+        $this->output = $output;
+
+        $this->router = Router::getInstance();
+    }
+
+    public function compileStaticPages(): void
+    {
+        $this->getDiscoveredModels()->each(function (string $pageClass) {
+            $this->compilePagesForClass($pageClass);
+        });
+    }
+
+    public function cleanOutputDirectory(): void
+    {
+        if (config('hyde.empty_output_directory', true)) {
+            $this->warn('Removing all files from build directory.');
+
+            if ($this->isItSafeToCleanOutputDirectory()) {
+                array_map('unlink', glob(Hyde::getSiteOutputPath('*.{html,json}'), GLOB_BRACE));
+                File::cleanDirectory(Hyde::getSiteOutputPath('media'));
+            }
+        }
+    }
+
+    public function transferMediaAssets(): void
+    {
+        $this->needsDirectory(Hyde::getSiteOutputPath('media'));
+
+        $collection = CollectionService::getMediaAssetFiles();
+        if ($this->canRunBuildAction($collection, 'Media Assets', 'Transferring')) {
+            $this->withProgressBar($collection,
+                function ($filepath) {
+                    copy($filepath, Hyde::getSiteOutputPath('media/'.basename($filepath)));
+                }
+            );
+            $this->newLine(2);
+        }
+    }
+
+    protected function getDiscoveredModels(): Collection
+    {
+        return $this->router->getRoutes()->map(function (Route $route) {
+            return $route->getPageType();
+        })->unique();
+    }
+
+    protected function canRunBuildAction(array|\Countable $collection, string $pageClass, ?string $verb = null): bool
```

Comment:
Using the router instead of the collection service brings a side effect; no empty collections will be passed to the build loop. While this sadly means we won't get the `No $name found. Skipping...` message, we can also reduce complexity by removing this helper method.

Reply:
Resolved in https://github.com/hydephp/develop/pull/174/commits/301b904888b87bc4155a9cdcd9263930be24070c

Reply:
Remember to update the changelog!